netket/graph/_lattice_edge_logic.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2022 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import numpy as np
+from scipy.spatial import cKDTree
+from scipy.sparse import find, triu
+from netket.utils.float import comparable, is_approx_int
+
+# Near-neighbour search logic
+
+
+def create_padded_sites(basis_vectors, extent, site_offsets, pbc, order):
+    """Generates all lattice sites in an extended shell"""
+    # note: by modifying these, the number of shells can be tuned.
+    shell_vec = np.where(pbc, 2 * order, 0)
+    shift_vec = np.where(pbc, order, 0)
+
+    shell_min = 0 - shift_vec
+    shell_max = np.asarray(extent) + shell_vec - shift_vec
+
+    # cell coordinates
+    ranges = [slice(lo, hi) for lo, hi in zip(shell_min, shell_max)]
+    # site coordinate within unit cell
+    ranges += [slice(0, len(site_offsets))]
+
+    basis_coords = np.mgrid[ranges].reshape(len(extent) + 1, -1).T
+    positions = basis_coords[:, :-1] @ basis_vectors
+    positions = positions.reshape(-1, len(site_offsets), len(extent)) + site_offsets
+    positions = positions.reshape(-1, len(extent))
+
+    basis_coords[:, :-1] %= extent
+    radix = np.cumprod([1, len(site_offsets), *extent[:0:-1]])[::-1]
+    equivalent_ids = basis_coords @ radix
```

Comment:
It would be good to have a comment here explaining how the generation of ids works here.